utp-testing/src/lib.rs

```diff
@@ -0,0 +1,184 @@
+extern crate core;
+
+pub mod cli;
+pub mod rpc;
+
+use crate::rpc::RpcServer;
+use discv5::{Discv5Event, TalkRequest};
+use jsonrpsee::core::{async_trait, RpcResult};
+use jsonrpsee::http_server::{HttpServerBuilder, HttpServerHandle};
+use jsonrpsee::proc_macros::rpc;
+use log::debug;
+use std::net::SocketAddr;
+use std::str::FromStr;
+use std::sync::Arc;
+use tokio::sync::mpsc::{UnboundedReceiver, UnboundedSender};
+use tokio::sync::RwLock;
+use trin_core::portalnet::discovery::Discovery;
+use trin_core::portalnet::types::messages::{PortalnetConfig, ProtocolId};
+use trin_core::portalnet::Enr;
+use trin_core::utils::bytes::hex_encode;
+use trin_core::utp::stream::{
+    UtpListener, UtpListenerEvent, UtpListenerRequest, UtpPayload, UtpStream,
+};
+use trin_core::utp::trin_helpers::UtpStreamId;
+
+/// uTP test app
+pub struct TestApp {
+    pub discovery: Arc<Discovery>,
+    pub utp_listener_tx: UnboundedSender<UtpListenerRequest>,
+    pub utp_listener_rx: Arc<RwLock<UnboundedReceiver<UtpListenerEvent>>>,
+    pub utp_event_tx: UnboundedSender<TalkRequest>,
+    pub utp_payload: Arc<RwLock<Vec<UtpPayload>>>,
+}
+
+#[async_trait]
+impl RpcServer for TestApp {
+    fn local_enr(&self) -> RpcResult<String> {
+        Ok(self.discovery.local_enr().to_base64())
+    }
+
+    async fn get_utp_payload(&self) -> RpcResult<String> {
+        let utp_payload = self.utp_payload.read().await;
+        let utp_payload = utp_payload.last();
+
+        match utp_payload {
+            Some(payload) => Ok(hex_encode(payload)),
+            None => Ok("None".to_string()),
+        }
+    }
+
+    async fn prepare_to_recv(&self, enr: String, conn_id: u16) -> RpcResult<String> {
+        let enr = Enr::from_str(&*enr).unwrap();
+        self.prepare_to_receive(enr, conn_id).await;
+        Ok("OK".to_owned())
```

Comment:
I'm not so excited about using `"OK"` and `"None"` as return types... It looks like `RpcResult<bool>` might be a more suitable type?